<a href="https://colab.research.google.com/github/pravallika-1305/WISE_ML/blob/main/Bagging_Boosting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier, BaggingClassifier, AdaBoostClassifier


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = pd.read_csv('/content/drive/My Drive/Data/mnist_train.csv')

In [4]:
df.head()

label  1x1  1x2  1x3  1x4  1x5  ...  28x23  28x24  28x25  28x26  28x27  28x28
0      5    0    0    0    0    0  ...      0      0      0      0      0      0
1      0    0    0    0    0    0  ...      0      0      0      0      0      0
2      4    0    0    0    0    0  ...      0      0      0      0      0      0
3      1    0    0    0    0    0  ...      0      0      0      0      0      0
4      9    0    0    0    0    0  ...      0      0      0      0      0      0

[5 rows x 785 columns]

The `mnist_train.csv` file contains the 60,000 training examples and labels.
<br>
The `mnist_test.csv` contains 10,000 test examples and labels.
<br>
Each row consists of 785 values:
<br>
1. the first value is the label (a number from 0 to 9)
2. the remaining 784 values are the pixel values (a number from 0 to 255).

In [5]:
df.shape

(60000, 785)

In [6]:
test = pd.read_csv('/content/drive/My Drive/Data/mnist_test.csv')

In [7]:
test.shape

(10000, 785)

In [8]:
x_train = df.iloc[:,1:]
y_train = df.iloc[:,0]

x_test = test.iloc[:,1:]
y_test = test.iloc[:,0]

## Random Forest

In [9]:
rf = RandomForestClassifier(n_estimators=10)
rf.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [10]:
rf.score(x_train,y_train)*100

99.91166666666666

In [11]:
rf.score(x_test,y_test)*100

94.93

## Bagging
Bagging is an abbreviation for "bootstrap aggregating".<br>
It'a meta-algorithm, which takes M subsamples (with replacement) from the initial dataset and trains the predictive model on those subsamples.<br> The final model is obtained by averaging the "bootstrapped" models.

In [12]:
bg = BaggingClassifier(DecisionTreeClassifier(),max_samples=0.5, max_features=1.0, n_estimators=20)
bg.fit(x_train,y_train)

BaggingClassifier(base_estimator=DecisionTreeClassifier(ccp_alpha=0.0,
                                                        class_weight=None,
                                                        criterion='gini',
                                                        max_depth=None,
                                                        max_features=None,
                                                        max_leaf_nodes=None,
                                                        min_impurity_decrease=0.0,
                                                        min_impurity_split=None,
                                                        min_samples_leaf=1,
                                                        min_samples_split=2,
                                                        min_weight_fraction_leaf=0.0,
                                                        presort='deprecated',
                                                        random_state=None,


In [13]:
bg.score(x_test,y_test)*100

94.76

In [14]:
bg.score(x_train,y_train) * 100

99.37333333333333

## Boosting
Boosting refers to a family of algorithms that are able to convert weak learners to strong learners. The main principle of boosting is to fit a sequence of weak learners such as small decision trees to weighted versions of the data. More weight is given to examples that were misclassified by earlier rounds. The predictions are then combined through a weighted majority vote (classification) or a weighted sum (regression) to produce the final prediction.

In [15]:
ad = AdaBoostClassifier(DecisionTreeClassifier(), n_estimators=10, learning_rate=0.01)
ad.fit(x_train,y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=DecisionTreeClassifier(ccp_alpha=0.0,
                                                         class_weight=None,
                                                         criterion='gini',
                                                         max_depth=None,
                                                         max_features=None,
                                                         max_leaf_nodes=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                                         min_weight_fraction_leaf=0.0,
                                                         presort='deprecated',
                       

In [16]:
ad.score(x_test,y_test)*100

87.63

In [17]:
ad.score(x_train,y_train)*100

100.0

Bagging decreases variance, while boosting decreases bias.
<br>
Also, to be noted that under-fitting means that the model has low variance and high bias and vice versa for overfitting.
<br> 
So, boosting is more vulnerable to overfitting than bagging.